modules/core/src/lib/deck.js

```diff
@@ -705,9 +705,14 @@ export default class Deck {
 
     this.props.onBeforeRender({gl});
 
+    const layers = this.layerManager.getLayers();
+    const drawLayers = layers.filter(l => l.props.operation === 'draw');
+    const maskLayers = layers.filter(l => l.props.operation === 'mask');
+    maskLayers.forEach(layer => layer.setNeedsRedraw());
```

Comment:
`setNeedsRedraw` should never be called outside of the layer lifecycle / layer manager.

Filtering logic is in passes/layers-pass.js. The Deck instance should not contain any specific knowledge of how to draw layers.